includes/Core/Authentication/Clients/OAuth_Client.php

```diff
@@ -532,6 +558,128 @@ function( $scope ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns the setup URL to the authentication proxy.
+	 *
+	 * @param string $code Optional. Temporary access code for an undelegated access token. Default empty string.
+	 * @return string URL to the setup page on the authentication proxy.
+	 */
+	public function get_proxy_setup_url( $code = '' ) {
+		$url = self::PROXY_URL . '/site-management/setup/';
+
+		$credentials = $this->get_client_credentials();
+
+		if ( ! is_object( $credentials ) || empty( $credentials->web->client_id ) ) {
+			$home_url           = home_url();
+			$home_url_no_scheme = str_replace( array( 'http://', 'https://' ), '', $home_url );
+
+			$rest_root = str_replace( array( 'http://', 'https://' ), '', rest_url() );
+			$rest_root = str_replace( $home_url_no_scheme, '', $rest_root );
+
+			$admin_root = str_replace( array( 'http://', 'https://' ), '', admin_url() );
+			$admin_root = str_replace( $home_url_no_scheme, '', $admin_root );
+
+			$nonce = $this->options->get( self::OPTION_PROXY_NONCE );
+			if ( empty( $nonce ) ) {
+				$nonce = wp_create_nonce( 'googlesitekit_proxy' );
+				$this->options->set( self::OPTION_PROXY_NONCE, $nonce );
+			}
+
+			return add_query_arg(
+				array(
+					'nonce'      => $nonce,
+					'name'       => rawurlencode( get_bloginfo( 'name' ) ),
+					'url'        => rawurlencode( $home_url ),
+					'rest_root'  => rawurlencode( $rest_root ),
+					'admin_root' => rawurlencode( $admin_root ),
+				),
+				$url
+			);
+		}
+
+		return add_query_arg(
+			array(
+				'site_id' => $credentials->web->client_id,
+				'code'    => $code,
+			),
+			$url
+		);
+	}
+
+	/**
+	 * Returns the permissions URL to the authentication proxy.
+	 *
+	 * This only returns a URL if the user already has an access token set.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return string URL to the permissions page on the authentication proxy on success,
+	 *                or empty string on failure.
+	 */
+	public function get_proxy_permissions_url() {
+		$access_token = $this->get_access_token();
+		if ( empty( $access_token ) ) {
+			return '';
+		}
+
+		$query_args = array( 'token' => $access_token );
+
+		$credentials = $this->get_client_credentials();
+		if ( is_object( $credentials ) && ! empty( $credentials->web->client_id ) ) {
+			$query_args['site_id'] = $credentials->web->client_id;
+		}
+
+		return add_query_arg(
+			$query_args,
+			self::PROXY_URL . '/site-management/permissions/'
+		);
+	}
+
+	/**
+	 * Checks whether the given proxy nonce is valid.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param string $nonce Nonce to validate.
+	 * @return bool True if nonce is valid, false otherwise.
+	 */
+	public function validate_proxy_nonce( $nonce ) {
+		$valid_nonce = $this->options->get( self::OPTION_PROXY_NONCE );
+		if ( $nonce !== $valid_nonce ) {
+			return false;
+		}
+
+		$this->options->delete( self::OPTION_PROXY_NONCE );
+		return true;
```

Comment:
> A typical WordPress nonce is user-dependent, so it would not work (because in the POST request nobody is considered logged-in, while on nonce generation a user is logged-in).

Yeah, that makes sense. 👍 

> Also WordPress nonces aren't real nonces, so arguably this approach here is more secure (because here it can really just be used once).

Yep 😄